iceberg/streams/iceberg/file_reader_builder.cpp

```diff
@@ -0,0 +1,126 @@
+#include "iceberg/streams/iceberg/file_reader_builder.h"
+
+namespace iceberg {
+
+namespace {
+
+void InsertOrFail(std::unordered_set<int>& set, int x) {
+  Ensure(!set.contains(x), "Field ids are not unique");
+  set.insert(x);
+}
+
+std::unordered_map<int, int> MakeIndexToFieldIdMap(const parquet::schema::GroupNode* node,
+                                                   const std::optional<std::string>& schema_name_mapping) {
+  if (!schema_name_mapping.has_value()) {
+    return {};
+  }
+  std::unordered_set<int> field_ids_set;
+  std::optional<SchemaNameMapper> schema_name_mapper;
+  std::optional<SchemaNameMapper::Node> root_node;
+  std::unordered_map<int, int> index_to_field_id;
+  for (int i = 0; i < node->field_count(); ++i) {
+    if (node->field(i)->field_id() >= 0) {
+      InsertOrFail(field_ids_set, node->field(i)->field_id());
+      continue;
+    }
+    if (!schema_name_mapper.has_value()) {
+      schema_name_mapper = SchemaNameMapper(*schema_name_mapping);
+      root_node = schema_name_mapper.value().GetRootNode();
+    }
+    auto field_name = node->field(i)->name();
+    auto field_id = root_node->GetFieldIdByName(field_name);
+    if (field_id.has_value()) {
+      index_to_field_id[i] = field_id.value();
+      InsertOrFail(field_ids_set, field_id.value());
+    }
+  }
+  return index_to_field_id;
+}
+
+}  // namespace
+
+std::vector<FileReaderBuilder::ParquetColumnInfo> FileReaderBuilder::GetColumnPositionsToRetrieveByFieldIds(
+    const parquet::FileMetaData& metadata, const std::vector<int>& field_ids) const {
+  const auto& schema = metadata.schema();
+  Ensure(schema != nullptr, std::string(__PRETTY_FUNCTION__) + ": schema is nullptr");
+
+  auto node = schema->group_node();
+
+  GroupNodeWrapper group_node_wrapper(node, MakeIndexToFieldIdMap(node, schema_name_mapping_));
+
+  const auto field_count = group_node_wrapper.GetFieldCount();
+
+  std::vector<ParquetColumnInfo> column_infos;
+
+  for (const auto& field_id_to_find : field_ids) {
+    bool found = false;
+
+    for (int i = 0; i < field_count; ++i) {
+      if (field_id_to_find == group_node_wrapper.GetRealFieldId(i)) {
+        found = true;
+        auto result_name = mapper_->FieldIdToColumnName(field_id_to_find);
+        column_infos.emplace_back(field_id_to_find, i, group_node_wrapper.GetName(i), std::move(result_name));
+        break;
+      }
+    }
+
+    if (!found) {
+      auto result_name = mapper_->FieldIdToColumnName(field_id_to_find);
```

Comment:
We call it in both logical branches. Let's extract before cycle to show that it's the same and would be called any case.